### PR TITLE
Use gpu picking for points, streamline/share picking code some more

### LINKED ISSUE #342: It's too hard to select points in the presence of lines
[https://user-images.githubusercontent.com/2624717/203812656-f5ef05dc-8655-4d44-8852-51021d2812e5.mov](https://user-images.githubusercontent.com/2624717/203812656-f5ef05dc-8655-4d44-8852-51021d2812e5.mov)

It feels unreasonably difficult to select points in the presence of lines. See video.

> From [SyncLinear.com](https://synclinear.com)

### PR DESCRIPTION
**Pull request by Wumpf**
### WHAT & HOW

* Picking of points (excluding depth_clouds!) is now done via GPU picking, just as meshes before in #1737 
* gpu picking now integrates seamless into other picking sources
   * gpu picking results combine now with other picking souces
   * introduces ui-space snapping of the picking target; as before anything that is close enough to the mouse cursor in a certain ui-point-radius will be picked!
   * some more cleanup to be done once more things are moved over to gpu based picking (follow-up PRs)
* finally fixed redundancy of picking handling between 2d & 3d scenes
   * more work is needed to have less parameters passed and handle other scene types as well (related to making more views re_renderer driven)

An interesting side effect is that we no longer need to keep around user data on point data. Which also opens the door to more writing directly to gpu memory!

### WHY

Towards #1615 

Point radius is now correctly taken into account when picking!

https://user-images.githubusercontent.com/1220815/231188920-7d007f5b-97eb-441a-bb23-69ded1f5e6c3.mov

Which finally fixes #342 (hard to select points in the presence of lines)

#1021 (image hover info should always show) is now fairly easy to fix, but keeping that for a follow-up.

### CODE REVIEW

File: crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs [hunk 1]
```diff
@@ -42,7 +42,8 @@ pub trait ScenePart {
 
 /// Computes the instance hash that should be used for picking (in turn for selecting/hover)
 ///
-/// Takes into account the currently the object properties, currently highlighted objects, and number of instances.
+/// TODO(andreas): Resolve the picking upon picking instead of doing it ahead of time here. This will speed up things.
```
Comment (teh-cmc) [inline crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs:45]: "Resolve the picking upon picking" :question:

File: crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs [hunk 2]
```diff
@@ -51,12 +52,40 @@ pub fn instance_path_hash_for_picking<T: re_log_types::Component>(
...
+///
+/// TODO(andreas): Resolve the picking upon picking instead of doing it ahead of time here. This will speed up things.
+/// Right now this is a bit hard to do since number of instances depends on the Primary. This is expected to change soon.
+pub fn instance_key_for_picking<T: re_log_types::Component>(
+    instance_key: re_log_types::component_types::InstanceKey,
+    entity_view: &re_query::EntityView<T>,
+    any_part_selected: bool,
+) -> re_log_types::component_types::InstanceKey {
+    if entity_view.num_instances() == 1 || !any_part_selected {
+        re_log_types::component_types::InstanceKey::SPLAT
```
Comment (teh-cmc) [inline crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs:76]: Some explanations would be welcome here, I have no idea what's going on?

An entity with a single instance isn't necessarily splat? But then again this logic predates this PR anyway.

**Review by teh-cmc — APPROVED** [review]
There sure is a lot happening ^_^

Worked fine in all my tests and generally makes sense to me.

File: crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs (hunk 1 above)
Comment (Wumpf) [reply crates/re_viewer/src/ui/view_spatial/scene/scene_part/mod.rs:45]: the ai is taking over!

File: crates/re_renderer/src/point_cloud_builder.rs [hunk 3]
```diff
@@ -357,21 +300,24 @@ where
...
     /// Pushes additional outline mask ids for this point
     ///
     /// Prefer the `overall_outline_mask_ids` setting to set the outline mask ids for the entire batch whenever possible!
+    #[inline]
     pub fn outline_mask_id(mut self, outline_mask_id: OutlineMaskPreference) -> Self {
         self.outline_mask_id = outline_mask_id;
         self
     }
+
+    /// This mustn't call this more than once.
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_renderer/src/point_cloud_builder.rs:312]: so turns out the entire interface has this stuff which is kinda bonkers (and my fault). Was thinking about this in relation to #1779 (crash on too many points). Need to overhaul this a bit. Will take in a followup

**Pull request merged**